### PR TITLE
linien: init at 1.0.0

### PR DESCRIPTION
**Pull request by fsagbuya**
## Description of changes

Add `linien` to nixpkgs. Structure needs to be checked since this have multiple sub-packages. Will comply to whatever is recommended.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/li/linien-client/package.nix
Comment (natsukium) [inline pkgs/by-name/li/linien-client/package.nix:9]: They share source code and need to be updated at the same time, I think.
In such cases, you can inherit version and src from `linien-common`.

e.g. https://github.com/NixOS/nixpkgs/blob/e4ad989506ec7d71f7302cc3067abd82730a4beb/pkgs/development/python-modules/openllm/default.nix#L52

File: pkgs/by-name/li/linien-client/package.nix
Comment (natsukium) [inline pkgs/by-name/li/linien-client/package.nix:33]:
```suggestion
    license = licenses.gpl3Plus;
```
https://github.com/linien-org/linien/blob/47f8c58477ff30f07147fbc98f251446365a95e3/README.md?plain=1#L455

File: pkgs/by-name/li/linien-client/package.nix
Comment (kirillrdy) [inline pkgs/by-name/li/linien-client/package.nix:12]: I haven't done anything like that with by-name, but is it possible to extract common ( repetitive bits ) into something like pkgs/by-name/li/linien-common/common.nix ?

File: pkgs/by-name/li/linien-common/package.nix
Comment (kirillrdy) [inline pkgs/by-name/li/linien-common/package.nix:7]: can linien-common be used as standalone package ? is there some binaries ? ( sorry I havent tried building this yet )

File: pkgs/by-name/li/linien-client/package.nix
Comment (natsukium) [inline pkgs/by-name/li/linien-client/package.nix:31]:
```suggestion
    description = "Client components of the Linien spectroscopy lock application";
```
Please remove the trailing period.
https://nixos.org/manual/nixpkgs/stable/#var-meta-description

File: pkgs/by-name/li/linien-client/package.nix
Comment (natsukium) [inline pkgs/by-name/li/linien-client/package.nix:10]:
```suggestion
  pyproject = true;
```

File: pkgs/by-name/li/linien-common/package.nix
Comment (natsukium) [inline pkgs/by-name/li/linien-common/package.nix:14]:
```suggestion
    rev = "v${version}";
```

File: pkgs/by-name/li/linien-client/package.nix
Comment (natsukium) [reply pkgs/by-name/li/linien-client/package.nix:12]: maybe?
```
 .
└──  pkgs
    └──  by-name
        └──  li
            └──  linien
                ├──  client.nix
                ├──  common.nix
                ├──  gui.nix
                ├──  server.nix
                └──  test.nix
```

Comment (kirillrdy) [reply pkgs/by-name/li/linien-client/package.nix:12]: i am not familiar with by-name loading code, will it only load `package.nix` ?

Comment (fsagbuya) [reply pkgs/by-name/li/linien-client/package.nix:12]: I can also transfer it to other than with pkgs/by-name. Not sure where to put it though

File: pkgs/by-name/li/linien-common/package.nix
Comment (fsagbuya) [reply pkgs/by-name/li/linien-common/package.nix:7]: There's a `setup.py` for the `linien-common`. I think it can be a standalone package.

File: pkgs/by-name/li/linien-client/package.nix
Comment (kirillrdy) [inline pkgs/by-name/li/linien-client/package.nix:27]: `maintainers.fsagbuya`

File: pkgs/by-name/li/linien-common/package.nix
Comment (kirillrdy) [reply pkgs/by-name/li/linien-common/package.nix:7]: buildPythonApplication is for applications and this is a library

File: pkgs/by-name/li/linien-client/package.nix
Comment (kirillrdy) [inline pkgs/by-name/li/linien-client/package.nix:6]: linien-client is also a library

File: pkgs/by-name/li/linien-gui/package.nix
Comment (kirillrdy) [inline pkgs/by-name/li/linien-gui/package.nix:7]: doesn't start
```
Traceback (most recent call last):
  File "/nix/store/8kvb3hhw9hvpzmjkp4i873f9rv11kz50-linien-gui-0.8.0/bin/.linien-wrapped", line 6, in <module>
    from linien_gui.app import run_application
  File "/nix/store/8kvb3hhw9hvpzmjkp4i873f9rv11kz50-linien-gui-0.8.0/lib/python3.11/site-packages/linien_gui/app.py", line 23, in <module>
    from linien_client.connection import LinienClient
  File "/nix/store/4hwmg9xyvn9k7hr4ij66pva6pyc60425-linien-client-0.8.0/lib/python3.11/site-packages/linien_client/connection.py", line 27, in <module>
    from linien_common.config import DEFAULT_SERVER_PORT
  File "/nix/store/iyzbhv1ll28798y91f075zk41nniqdhr-linien-common-0.8.0/lib/python3.11/site-packages/linien_common/config.py", line 21, in <module>
    from appdirs import AppDirs
ModuleNotFoundError: No module named 'appdirs'
```

File: pkgs/by-name/li/linien-server/package.nix
Comment (kirillrdy) [inline pkgs/by-name/li/linien-server/package.nix:6]: ./result/bin/linien-server 

```
  File "/nix/store/38aa8gm3nyssciinn10x9ym6c5gmmw4c-linien-server-0.8.0/lib/python3.11/site-packages/linien_server/registers.py", line 52, in __init__
    from linien_server.acquisition import AcquisitionService
  File "/nix/store/38aa8gm3nyssciinn10x9ym6c5gmmw4c-linien-server-0.8.0/lib/python3.11/site-packages/linien_server/acquisition.py", line 32, in <module>
    from pyrp3.board import RedPitaya  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'pyrp3'
Saved parameters to  /home/kirillvr/.local/share/linien/linien_parameters.json
```

**Review by kirillrdy — COMMENTED** [review]
overall, i think there no point of exposing python libraries in pkgs/by-name since they wont be able to be used in python.pkgs

this is a bit of a new ground for me, ( and i assume many others )

some options:

- move library bits into python-modules (  either import them in apps here, or move everything there and use toPythonApplication in pkgs/by-name
- only expose applications in pkgs/by-name

File: pkgs/by-name/li/linien-common/package.nix
Comment (natsukium) [reply pkgs/by-name/li/linien-common/package.nix:7]: I guess it would be better to move linien-common to python-modules and leave the other three in pkgs/by-name.

File: pkgs/by-name/li/linien-client/package.nix
Comment (natsukium) [reply pkgs/by-name/li/linien-client/package.nix:6]: I think linien-client is an application that has the command linien.
https://github.com/linien-org/linien#installation-with-pip

Comment (kirillrdy) [reply pkgs/by-name/li/linien-client/package.nix:6]:
```
nix-build -A linien-client
/nix/store/4hwmg9xyvn9k7hr4ij66pva6pyc60425-linien-client-0.8.0


ls ./result/
lib  nix-support
```

Comment (natsukium) [reply pkgs/by-name/li/linien-client/package.nix:6]: Indeed, they don't provide scripts.
https://github.com/search?q=repo%3Alinien-org%2Flinien+entry_points&type=code

Comment (fsagbuya) [reply pkgs/by-name/li/linien-client/package.nix:6]: Will move client and common to python-modules and retain server and gui in pkgs/by-name

File: pkgs/by-name/li/linien-gui/package.nix
Comment (fsagbuya) [reply pkgs/by-name/li/linien-gui/package.nix:5]: added `appdirs` to propagatedBuildInputs. Though it stops again,
```
./result/bin/linien
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

fish: Job 1, './result/bin/linien' terminated by signal SIGABRT (Abort)
```

**Comment by fsagbuya** [comment]
Why it errors at this?
```
pkgs/top-level/python-packages.nix:
	6368: Trailing whitespace

1 errors found
Error: Process completed with exit code 123.
```
I already double checked that line and there's no trailing whitespace.

File: pkgs/by-name/li/linien-server/package.nix
Comment (fsagbuya) [reply pkgs/by-name/li/linien-server/package.nix:5]: Should I package `pyrp3` too?

File: pkgs/development/python-modules/linien-common/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/linien-common/default.nix:33]: Whats with tests and the pythonImportsCheck?

File: pkgs/development/python-modules/linien-client/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/linien-client/default.nix:26]: Whats with tests and the pythonImportsCheck?

File: pkgs/by-name/li/linien-server/package.nix
Comment (SuperSandro2000) [reply pkgs/by-name/li/linien-server/package.nix:5]: If the software has a hard dependency on it, then probably yes

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/by-name/li/linien-server/package.nix
Comment (fsagbuya) [reply pkgs/by-name/li/linien-server/package.nix:5]: Running now with:
```
./result/bin/linien-server --fake
```
https://github.com/linien-org/linien/wiki/Development#optional-install-fake-server

File: pkgs/development/python-modules/linien-client/default.nix
Comment (fsagbuya) [reply pkgs/development/python-modules/linien-client/default.nix:26]: `pythonImportsCheck` added

File: pkgs/development/python-modules/linien-common/default.nix
Comment (fsagbuya) [reply pkgs/development/python-modules/linien-common/default.nix:33]: `pythonImportsCheck` added

File: pkgs/by-name/li/linien-gui/package.nix
Comment (fsagbuya) [reply pkgs/by-name/li/linien-gui/package.nix:6]: 'appdirs' is now added

File: pkgs/development/python-modules/linien-common/default.nix
Comment (natsukium) [reply pkgs/development/python-modules/linien-common/default.nix:33]: I think we can add [these tests](https://github.com/linien-org/linien/tree/master/tests) like ipykernel. 
https://github.com/NixOS/nixpkgs/tree/nixos-unstable/pkgs/development/python-modules/ipykernel

Comment (fsagbuya) [reply pkgs/development/python-modules/linien-common/default.nix:33]: Hmm can you help provide more details how to create the `tests.nix` file here? Like what dependencies I should use.

Comment (fsagbuya) [reply pkgs/development/python-modules/linien-common/default.nix:33]: @natsukium

Comment (natsukium) [reply pkgs/development/python-modules/linien-common/default.nix:33]: Sorry for the late reply.

We need to add pytestCheckHook, linien-common and linien-client to nativeCheckInputs.
Also, looking at some test files, it looks like numpy and scipy are needed as well.

Regarding the test packages, ipykernel's tests.nix would be helpful, and you can use buildPythonPackages with `pyproject = false`.
The src could be shared with linien-common.

Comment (fsagbuya) [reply pkgs/development/python-modules/linien-common/default.nix:34]: @natsukium updated

File: pkgs/development/python-modules/linien-common/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/linien-common/default.nix:14]: Could you please update to version v1.0.0 ?

https://github.com/linien-org/linien/releases

**Review by doronbehar — CHANGES_REQUESTED** [review]
Thanks for calling me for a review. I would like to be very involved in the addition of this package, but I see various issues. I also wonder how come many dependecies I noticed needed when I first wanted to add these packages in https://github.com/NixOS/nixpkgs/pull/195144 weren't required. In anycase, I barely have time nowadays :cry: , so I'm afraid it will take me time to approve this after the next iteration.

Here are some general comments I have:

What would be the purpose of adding the `linien-server` package? It can run only on a RedPitaya OS - `armv7l-linux`, and Nix, nor NixOS don't support that yet. Not only that, it doesn't support running `linien-server` as a systemd service. I talked with upstream about this issues in numerous threads in their bug tracker. I will also note that I managed to add a `flake.nix` to linien's repo, and to cross compile it to armv7l, and I even managed to patch it to support a systemd based setup - that's all mentioned in these issues:

https://github.com/linien-org/linien/issues?q=is%3Aissue+is%3Aopen+involves%3Adoronbehar

Due to all of that, I'd remove linien-server from this PR in the meantime.

Additionally, the v1.0.0 update is supposed to be important. Please make sure that at least `linien-gui` v1.0.0 works with a real redpitaya running a linien-server v1.0.0.

File: pkgs/development/python-modules/linien-common/default.nix
Comment (doronbehar) [reply pkgs/development/python-modules/linien-common/default.nix:34]: > This is because this repository has a monorepo like structure. linien-common, linien-client, ... packages are in the subdirectory, but tests is defined at the top level.

Thanks for the explanation.

Marked as resolved.

File: pkgs/development/python-modules/linien-common/default.nix
Comment (doronbehar) [reply pkgs/development/python-modules/linien-common/default.nix:14]: The issue you showed is fixed in many places in Nixpkgs with:

```nix
preCheck = ''
  export HOME=$(mktemp -d)
'';
```

Comment (fsagbuya) [reply pkgs/development/python-modules/linien-common/default.nix:14]: Issue still occurs after adding the `preCheck`.

**Comment by doronbehar** [comment]
> I only attempt to build with the minimal requirements base from `setup.py`. If there's any specific dependencies I might have missed, you are very welcome to suggest.

Fair enough. Perhaps things changed upstream.

> I will remove it for now if this is the case.

Yes please - it doesn't make sense to add a package that we cannot even test, and doesn't fit running on NixOS.

> Unfortunately, I don't have red pitaya to test this. I can only test on the host platform if I can start the gui. If you have the hardware, you are welcome to test it and suggest changes.

That would be ideal, I'll update you if I get the chance to do so.

**Comment by doronbehar** [comment]
```
(ins)[nix-shell:~/.cache/nixpkgs-review/pr-269358-1]$ ./results/linien-gui/bin/linien
Traceback (most recent call last):
  File "/nix/store/kcx6gnm85y267954i0a05r90g4n0cahd-python3.10-numpy-1.26.1/lib/python3.10/site-packages/numpy/core/__init__.py", line 24, in <module>
    from . import multiarray
  File "/nix/store/kcx6gnm85y267954i0a05r90g4n0cahd-python3.10-numpy-1.26.1/lib/python3.10/site-packages/numpy/core/multiarray.py", line 10, in <module>
    from . import overrides
  File "/nix/store/kcx6gnm85y267954i0a05r90g4n0cahd-python3.10-numpy-1.26.1/lib/python3.10/site-packages/numpy/core/overrides.py", line 8, in <module>
    from numpy.core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/nix/store/kcx6gnm85y267954i0a05r90g4n0cahd-python3.10-numpy-1.26.1/lib/python3.10/site-packages/numpy/__init__.py", line 130, in <module>
    from numpy.__config__ import show as show_config
  File "/nix/store/kcx6gnm85y267954i0a05r90g4n0cahd-python3.10-numpy-1.26.1/lib/python3.10/site-packages/numpy/__config__.py", line 4, in <module>
    from numpy.core._multiarray_umath import (
  File "/nix/store/kcx6gnm85y267954i0a05r90g4n0cahd-python3.10-numpy-1.26.1/lib/python3.10/site-packages/numpy/core/__init__.py", line 50, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.11 from "/nix/store/qp5zys77biz7imbk6yy85q5pdv7qk84j-python3-3.11.6/bin/python3.11"
  * The NumPy version is: "1.26.1"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/nix/store/v6g3if1sxmwyf0c59dh8nb2xcn35khx6-linien-gui-1.0.0/bin/.linien-wrapped", line 6, in <module>
    from linien_gui.app import run_application
  File "/nix/store/v6g3if1sxmwyf0c59dh8nb2xcn35khx6-linien-gui-1.0.0/lib/python3.11/site-packages/linien_gui/app.py", line 26, in <module>
    from linien_gui.ui.device_manager import DeviceManager
  File "/nix/store/v6g3if1sxmwyf0c59dh8nb2xcn35khx6-linien-gui-1.0.0/lib/python3.11/site-packages/linien_gui/ui/device_manager.py", line 21, in <module>
    from linien_gui.dialogs import (
  File "/nix/store/v6g3if1sxmwyf0c59dh8nb2xcn35khx6-linien-gui-1.0.0/lib/python3.11/site-packages/linien_gui/dialogs.py", line 29, in <module>
    from pyqtgraph import QtCore
  File "/nix/store/qfg5vl5c7dbkp27hslaj307m36mk8x0x-python3.11-pyqtgraph-0.13.3/lib/python3.11/site-packages/pyqtgraph/__init__.py", line 14, in <module>
    import numpy  # # pyqtgraph requires numpy
    ^^^^^^^^^^^^
  File "/nix/store/kcx6gnm85y267954i0a05r90g4n0cahd-python3.10-numpy-1.26.1/lib/python3.10/site-packages/numpy/__init__.py", line 135, in <module>
    raise ImportError(msg) from e
ImportError: Error importing numpy: you should not try to import numpy from
        its source directory; please exit the numpy source tree, and relaunch
        your python interpreter from there.
```

**Comment by fsagbuya** [comment]
@doronbehar I'm not able to reproduce the error.
```
nix-build -A linien-gui && ./result/bin/linien
/nix/store/qzkvk2w4laj5cvaf84pkmgz627gkbg56-linien-gui-1.0.0
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
linien_gui.app                 INFO     Starting Linien GUI
```

**Comment by doronbehar** [comment]
@fsagbuya it is possible you don't experience the issue because your branch is not rebased onto latest `master`. What `nixpkgs-review` does is to merge the changes ontop of current `master` and then build the new and modified packages of the PR. There was a rather recent `python-updates` branch merge that might have caused the issue I experienced.

Do the rebase using:

```
git checkout master
git pull
git checkout linien
git rebase master
```

**Comment by fsagbuya** [comment]
@doronbehar now I have rebase onto latest master and tried to build and it was successful. I wonder if some latest `python-updates` fixed the previous issue?
```
Rewriting #!/nix/store/qp5zys77biz7imbk6yy85q5pdv7qk84j-python3-3.11.6/bin/python3.11 to #!/nix/store/qp5zys77biz7imbk6yy85q5pdv7qk84j-python3-3.11.6
wrapping `/nix/store/v6g3if1sxmwyf0c59dh8nb2xcn35khx6-linien-gui-1.0.0/bin/linien'...
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
running install tests
no Makefile or custom installCheckPhase, doing nothing
pythonCatchConflictsPhase
pythonRemoveBinBytecodePhase
pythonImportsCheckPhase
Executing pythonImportsCheckPhase
postPatchMkspecs
/nix/store/v6g3if1sxmwyf0c59dh8nb2xcn35khx6-linien-gui-1.0.0
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
linien_gui.app                 INFO     Starting Linien GUI
```

File: pkgs/by-name/li/linien-gui/package.nix
Comment (doronbehar) [inline pkgs/by-name/li/linien-gui/package.nix:18]:
```suggestion
  ];
  
  # Makes qt-wayland appear in the qt paths injected by the wrapper - helps users
  # with `QT_QPA_PLATFORM=wayland` in their environment.
  buildInputs = [
    qt5.qtwayland
  ];
```

**Review by doronbehar — CHANGES_REQUESTED** [review]
Weird, the issue I observed was because the `nixpkgs-review` shell injected many Python3.10 paths to `$PYTHONPATH` and these in turn created import issues with `linien-gui` that uses Python3.11.

GUI seems to be functional, so this should be almost good to go. See the last code suggestion I gave.

**Review by doronbehar — APPROVED** [review]

**Pull request merged**
